core/trino-spi/src/main/java/io/trino/spi/Page.java

```diff
@@ -315,6 +315,21 @@ public Page getLoadedPage(int... columns)
         return wrapBlocksWithoutCopy(positionCount, blocks);
     }
 
+    public Page getLoadedPage(int[] columns, int[] eagerlyLoadedColumns)
+    {
+        requireNonNull(columns, "columns is null");
+
+        for (int column : eagerlyLoadedColumns) {
```

Comment:
If any block was actually loaded, you should also call `io.trino.spi.Page#updateRetainedSize`
Test in `io.trino.spi.TestPage` would be great

Reply:
The behaviour we have now seems to be consistent with the existing getLoadedPage methods.
E.g.
```
LazyBlock lazyBlock = lazyWrapper(block);
Page page = new Page(lazyBlock);
long lazyPageRetainedSize = Page.INSTANCE_SIZE + sizeOf(new Block[] {block}) + lazyBlock.getRetainedSizeInBytes();
assertEquals(page.getRetainedSizeInBytes(), lazyPageRetainedSize);
Page loadedPage = page.getLoadedPage();
assertEquals(page.getRetainedSizeInBytes(), lazyPageRetainedSize);
```
After the lazy block has been loaded its retained size has changed internally but the original Page which contains a loaded lazy block now doesn't have any change in retained size.
I'm wondering if there is a need to update the retained size of original page as `ProjectSelectedPositions#updateRetainedSize` appears to be doing its own accounting for retained size without relying on `Page#getRetainedSizeInBytes`.

Reply:
> The behaviour we have now seems to be consistent with the existing getLoadedPage methods.

`getLoadedPage` does not change current `Page` blocks (does not unwrap them)

Also, I'm not saying other methods are fully correct (they aren't when it comes to updating retained size)

Reply:
> getLoadedPage does not change current Page blocks (does not unwrap them)

Unwrapping or not shouldn't matter w.r.t accounting retained size. The page still has reference to a lazy block which after loading retains the updated size, not the old size. In fact, without unwrapping the retained size is larger due to the size of the wrapper.
Since this method is going to get called a lot, adding the overhead for memory accounting needs to be justified by its usage somewhere. In the current usage when this is called in PageProcessor, `retainedSizeInBytes` of the Page is still at -1. So any subsequent code calling `Page#getRetainedSizeInBytes` will already get the updated retained size, we don't need to do that update eagerly.

Reply:
> In the current usage when this is called in PageProcessor, retainedSizeInBytes of the Page is still at -1.

How do you know that? If that's the case that is a coincidence. You could:
1. check if `retainedSizeInBytes == -1` and skip update if that's the case
2. check if loaded block assignments actually changed (`Page#getLoadedPage()` does something like that) and skip update if nothing changed.

Reply:
I had checked a few queries and seen that all the code usages of `Page#getRetainedSizeInBytes` appear to be after PageProcessor projection (MergePages seems to be the nearest user of it).
I've added `if (retainedSizeInBytes != -1)  updateRetainedSize();` here now since that appears to be free in current code.